x/bsonx/bsoncore/document.go

```diff
@@ -264,58 +261,73 @@ func (d Document) DebugString() string {
 // String outputs an ExtendedJSON version of Document. If the document is not valid, this method
 // returns an empty string.
 func (d Document) String() string {
-	return d.StringN(math.MaxInt)
+	str, _ := d.stringN(0)
+	return str
 }
 
 // StringN stringifies a document upto N bytes
-func (d Document) StringN(n int) string {
-	if len(d) < 5 || n <= 0 {
-		return ""
+func (d Document) StringN(n int) (string, bool) {
+	if n <= 0 {
+		if l, _, ok := ReadLength(d); !ok || l < 5 {
+			return "", false
+		}
+		return "", true
 	}
+	return d.stringN(n)
```

Comment:
It's interesting that calling `StringN(0)` will return an empty string but `stringN(0)` will return the entire document. The asymmetry makes the code fairly difficult to read. Should we update `StringN` so that an N <= 0 argument returns the full document?

```go
 func (d Document) String(n int) (string, bool) {
	if n <= 0 {
		return d.stringN(0)
	}

	return d.stringN(n)
}
```

Ultimately, it's unclear to me why one would call `StringN(0)` expecting an empty string.

Reply:
By using `stringN()`, we can keep the original behavior of `StringN(n)`, which is to return an empty string when n <= 0, while also allowing it to be reused by `String()`. I will be happy to merge `stringN()` and `StringN()` if we all accept returning the entire document on `StringN(0)`.

Reply:
I have a slight preference for -1 (i.e. all negative) vs 0 (i.e. all non-positive) because it more closely matches the semantics of string truncation (i.e. `StringN(0) == String()[:0]`), although there's not a clear use case for passing either -1 or 0. Is there a reason we want to change it from the original where we pass `math.IntMax` to get the full document?

Reply:
-1 (i.e. all negative) vs 0 (i.e. all non-positive) seems a more sensible interface.

It's just for a better compatibility that not using a particular value like `math.IntMax` to get the entire document.